vault/plugin_reload.go

```diff
@@ -70,52 +70,103 @@ func (c *Core) reloadMatchingPluginMounts(ctx context.Context, mounts []string)
 	return errors
 }
 
-// reloadPlugin reloads all mounted backends that are of
-// plugin pluginName (name of the plugin as registered in
-// the plugin catalog).
-func (c *Core) reloadMatchingPlugin(ctx context.Context, pluginName string) error {
-	c.mountsLock.RLock()
-	defer c.mountsLock.RUnlock()
-	c.authLock.RLock()
-	defer c.authLock.RUnlock()
+// reloadMatchingPlugin reloads all mounted backends that are named pluginName
+// (name of the plugin as registered in the plugin catalog). It returns the
+// number of plugins that were reloaded and an error if any.
+func (c *Core) reloadMatchingPlugin(ctx context.Context, pluginName string) (reloaded int, err error) {
+	typeExists := map[consts.PluginType]bool{
+		consts.PluginTypeCredential: false,
+		consts.PluginTypeDatabase:   false,
+		consts.PluginTypeSecrets:    false,
+	}
+	for pluginType := range typeExists {
+		plugins, err := c.pluginCatalog.ListVersionedPlugins(ctx, pluginType)
+		if err != nil {
+			return reloaded, err
+		}
+		for _, plugin := range plugins {
+			if plugin.Name == pluginName {
+				typeExists[pluginType] = true
+				break
+			}
+		}
+	}
 
 	ns, err := namespace.FromContext(ctx)
 	if err != nil {
-		return err
+		return reloaded, err
 	}
 
-	// Filter mount entries that only matches the plugin name
-	for _, entry := range c.mounts.Entries {
-		// We dont reload mounts that are not in the same namespace
-		if ns.ID != entry.Namespace().ID {
-			continue
-		}
-		if entry.Type == pluginName || (entry.Type == "plugin" && entry.Config.PluginName == pluginName) {
-			err := c.reloadBackendCommon(ctx, entry, false)
-			if err != nil {
-				return err
+	// The combined database plugin itself is a secrets plugin, so we need to check
+	// the secrets mount table for database plugins.
+	if typeExists[consts.PluginTypeSecrets] || typeExists[consts.PluginTypeDatabase] {
+		c.mountsLock.RLock()
+		defer c.mountsLock.RUnlock()
+
+		for _, entry := range c.mounts.Entries {
+			// We dont reload mounts that are not in the same namespace
+			if ns.ID != entry.Namespace().ID {
+				continue
+			}
+
+			if entry.Type == pluginName || (entry.Type == "plugin" && entry.Config.PluginName == pluginName) {
+				err := c.reloadBackendCommon(ctx, entry, false)
+				if err != nil {
+					return reloaded, err
+				}
+				reloaded++
+				c.logger.Info("successfully reloaded plugin", "plugin", pluginName, "namespace", entry.Namespace(), "path", entry.Path, "version", entry.Version)
+			}
+
+			// Knowledge of whether a database plugin is in use within a particular
+			// database mount is internal to the database plugin, so we delegate
+			// the reload request with an internally routed request
+			if typeExists[consts.PluginTypeDatabase] && entry.Type == "database" {
```

Comment:
I wonder if this could/should be an `else if`?

Reply:
If I was designing the API from scratch, I would definitely say yes, but I left it this way for consistency with the way auth and secrets plugins are handled. For example, if someone had registered auth, secrets, and database plugins all called "my-plugin", putting `else if` here would mean the auth and the secrets instances get reloaded but not the database instances. In reality though, I expect it to be very rare for someone to have the same plugin name in use with 3 different types, so I also don't think it matters too much either way, the consistency argument just slightly edges it IMO. Happy to revisit this though if you disagree or have other considerations I haven't covered.

Reply:
Hmm I've just been re-reading and I think I see what you were getting at now. It's impossible for both conditions to be meaningfully satisfied because of the `entry.Type` constraints - so making it `else if` is functionally identical but clearer to read. I'll fix that up.